Configure ts-node via tsconfig

cspotcode opened the pull request:
Closes #4.  Still WIP, but I still welcome early code review.

cspotcode commented:
@blakeembrey @G-Rath  This is ready for a review.  It still needs docs updates but I'd rather write them after the code is reviewed.

blakeembrey commented on package.json:
```diff
@@ -11,11 +11,17 @@
+    "clean": "rimraf dist && rimraf tsconfig.schema.json && rimraf tsconfig.schemastore-schema.json",
+    "build": "npm run clean && npm run build:tsc && npm run build:configSchema",
+    "build:tsc": "tsc",
+    "build:configSchema": "typescript-json-schema --topRef --refs --validationKeywords allOf --out tsconfig.schema.json tsconfig.json TsConfigSchema && node --require ./register ./scripts/create-merged-schema",
```
Nit: Consistent dash casing is currently used.

blakeembrey commented on scripts/create-merged-schema.ts:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env ts-node
```
I'm assuming this line doesn't work, since we use `node --require` in the script?

cspotcode replied:
It should work if you have `ts-node` installed globally, but that's not technically true all the time, so I put `node --require` in the package script.

blakeembrey commented on scripts/create-merged-schema.ts:
```diff
@@ -0,0 +1,58 @@
+ */
+
+import axios from 'axios';
+import * as Path from 'path';
```
Nit: Casing on `Path` should be `path`, but you could just import `resolve` only.

blakeembrey commented on scripts/create-merged-schema.ts:
```diff
@@ -0,0 +1,58 @@
+      },
+    },
+    allOf: [
+      ...schemastoreSchema.allOf.slice(0, 4),
```
What does this mean? Why 4?

cspotcode replied:
Just so it gets spliced into the array at a point that looks nice.  https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/tsconfig.json#L630-L631

blakeembrey commented on scripts/create-merged-schema.ts:
```diff
@@ -0,0 +1,58 @@
+ * by merging the schemastore schema with our ts-node additions.
+ * This merged schema can be submitted in a pull request to
+ * SchemaStore.
+ */
```
Is this script necessary? Why can't we specify a URL within our schema to have it already be up-to-date with the remote schema? I'm worried this will mean that every TypeScript version I'll need to be generating and releasing the updated schema. You're meant to be able to use `$ref` for this use-case: https://cswr.github.io/JsonSchema/spec/definitions_references/.

cspotcode replied:
Correct, I'm generating 2 schemas.  First is `tsconfig.schema.json`, which *does* use `$ref`.  That was my initial approach, same thinking as you.

@G-Rath suggested patching ts-node's additions into the official file on Schemastore.  This script produces that merged schema.  The idea is that it would live in SchemaStore, so editors would tab-complete the ts-node fields without any configuration on the user's part.  I'm not sure if they'll ever merge it, though, because if someone tab-completes the ts-node options, they'll still need to install ts-node.  It could be nice publicity for the library; I dunno.

G-Rath replied:
They will merge it, since it's not a breaking change :)

The configuration options are harmless if `ts-node` isn't there.

blakeembrey commented on src/bin.ts:
```diff
@@ -79,29 +79,71 @@ export function main (argv: string[]) {
-    '--log-error': logError = DEFAULTS.logError,
-    '--emit': emit = DEFAULTS.emit
-  } = args
+  const flagOptions = {
```
I'm _really_ not a fan of these changes, it's super opaque and confusing. This is a large reason I said we should probably change the ordering in a follow up PR, it's taking away from the actual feature. In general, I think we should revert all these changes and tackle it in a follow up PR. If you still want to tackle it, move `DEFAULTS` into `index.ts` _only_ and don't use them in `bin`. The current code would be better represented by just extending `options, tsconfig, DEFAULTS` where each option is previously `undefined`.

cspotcode replied:
I'll roll them back.  If I understand you correctly, the implementation you're proposing is to never pass defaults from bin.ts into `register()` nor from `register()` into `create()`?  Instead, read them only when actually used?

blakeembrey replied:
Right, I believe we can read them only in `create()` if we use the `defaults()` function you created. This is because the values are all from `process.env` and should be undefined (or `null` in the case of `yn` being used).

cspotcode replied:
I'm upgrading yn to the latest version, which returns `undefined` instead of `null`.  This is good cuz it matches JS semantics for arg and destructuring defaults.

blakeembrey commented on src/bin.ts:
```diff
@@ -79,29 +79,71 @@ export function main (argv: string[]) {
+    requires: [],
+    code: undefined,
+    print: false,
+    interactive: false
```
None of these options should be passed to `register`, they're all CLI only.

blakeembrey commented on src/index.ts:
```diff
@@ -99,6 +158,81 @@ export interface RegisterOptions extends CreateOptions {
 }
 
+/*
+ * This interface exists solely for generating a JSON schema for tsconfig.json.
```
Can we put this in another file if it's not used here?

blakeembrey commented on src/index.ts:
```diff
@@ -99,6 +158,81 @@ export interface RegisterOptions extends CreateOptions {
+    | 'project'
+    | 'dir'
+  > {
+  requires?: Array<string>
```
Why this this allowed? It's not used except via CLI. I think we should not include any option support that isn't part of `register`.

cspotcode replied:
Yeah, it could be confusing.  I was thinking there might be a situation where a project bootstraps the environment by convention, maybe to tweak the source-map-support or something, so they put ts-node in the shebang and expect it to `require()` a bootstrapper.  I guess I was thinking that ts-node is, itself, bootstrapping the node environment to do something special, so it makes sense to allow that bootstrapping to be extended.

I can remove it if it's a blocker.

blakeembrey commented on src/index.ts:
```diff
@@ -99,6 +158,81 @@ export interface RegisterOptions extends CreateOptions {
+ * Helper for obtaining ts-node options from multiple sources and merging them.
+ * @internal
+ */
+export class OptionsHelper {
```
Remove this helper, just use `defaults()` in `register`.

blakeembrey commented on src/index.ts:
```diff
@@ -219,9 +355,20 @@ function cachedLookup <T> (fn: (arg: string) => T): (arg: string) => T {
  */
 export function register (opts: RegisterOptions = {}): Register {
-  const options = { ...DEFAULTS, ...opts }
+  return registerInternal(new OptionsHelper({
```
Can we keep only a single `register` function - not clear why we need a layer of indirection. If we implement options clearer it should be resolved.

blakeembrey commented on src/index.ts:
```diff
@@ -242,25 +389,55 @@ export function register (opts: RegisterOptions = {}): Register {
 export function create (options: CreateOptions = {}): Register {
+  return createInternal(new OptionsHelper({
+    explicitOptions: options
+  }))
```
Ditto here, remove the extra layer of indirection.

blakeembrey commented on src/index.ts:
```diff
@@ -647,9 +832,29 @@ function readConfig (
+  const tsconfigOptions: TsConfigOptions = {
+    ...config['ts-node']
+  }
+  if (tsconfigOptions.requires) {
```
If we want this we should do it as part of a follow up PR.

cspotcode replied:
Referring specifically to specifying `requires` in a tsconfig file?

blakeembrey commented on src/index.ts:
```diff
@@ -242,25 +389,55 @@ export function register (opts: RegisterOptions = {}): Register {
+  }
+
+  // compute enough options to read the config file
+  let { cwd, isScoped, compiler, ts, fileExists, readFile } = recomputedOptions()
```
I don't think we should be doing `recomputedOptions` like this, it feels wasteful. There's only two options here (`isScoped` and `compiler`) that can change from `tsconfig.json`. `isScoped` could be populated _after_ `tsconfig.json` is loaded, therefore this is really only here to load `compiler` again.

cspotcode replied:
I'll move `cwd` and `isScoped` outside of this function.

The other 4 -- `compiler`, `ts`, `readFile`, and `fileExists` -- are all needed both before and after `readConfig()`.  Should I rename the function to something like `loadCompiler`?  I can also get rid of the destructuring if you want; just assign straight to local `let` variables.

blakeembrey replied:
Yeah, for `readFile` and `fileExists`, we should probably just fallback on `ts.sys` within the `readConfig` function directly - I don't think those two functions are actually used consistently when I looked at the code, so it wouldn't be a loss to lose the functionality altogether.

cspotcode replied:
I see what you mean.  `ts.parseJsonConfigFileContent` accepts a `ParseConfigHost`, which has `readFile`, `fileExists`, and `readDirectory`.  We can add a `readDirectory` option to `RegisterOptions`, and then we can consistently use `readFile` and `fileExists` in a way that makes sense.

Who is using these options, and what do they use them for?

cspotcode replied:
It might be cleaner if ts-node accepted a `*Host`-style object instead of `readFile` and `fileExists`.  That way consumers of the API could override any of the various fs-related methods on `ts.sys`.  Looks like there are a bunch, like `directoryExists`, that are being used by ts-node but not overridable via the API.

blakeembrey commented on src/index.ts:
```diff
@@ -242,25 +389,55 @@ export function register (opts: RegisterOptions = {}): Register {
+  }
+
+  // Merge default options, tsconfig options, and explicit --flag options
+  options = optionsHelper.merge()
```
Use `defaults` directly, there's a layer of indirection and mutation here that's not necessary from what I can tell.

blakeembrey commented on tsconfig.json:
```diff
@@ -1,4 +1,6 @@
 {
+  "$schema": "./tsconfig.schemastore-schema.json",
+  "ts-node": {},
```
Is it empty for a reason?

cspotcode replied:
No reason, I forgot I added this.  I'll remove it.

blakeembrey commented on src/index.ts:
```diff
@@ -242,25 +389,55 @@ export function register (opts: RegisterOptions = {}): Register {
+  let { cwd, isScoped, compiler, ts, fileExists, readFile } = recomputedOptions()
+
+  // Read config file
+  const { config, options: tsconfigOptions } = readConfig(cwd, ts, fileExists, readFile, options)
```
We could be explicit about `options` being passed to `readConfig` here. One feature I think is broken due to the current design is the `files` option, that functionality could be moved into this function instead of `readConfig`.

cspotcode replied:
Good catch.  I think we need to set empty arrays for `files` and `includes` *within* `readConfig` because otherwise TypeScript's `parseJsonConfigFileContent` will do a bunch of globbing filesystem calls to discover the full list of files.  Does that sound right?

blakeembrey commented on src/index.ts:
```diff
@@ -99,6 +158,81 @@ export interface RegisterOptions extends CreateOptions {
+  for (const source of sources) {
+    for (const key of Object.keys(source)) {
+      const value = (source as any)[key]
+      if (value !== undefined) merged[key] = value
```
Probably want to do `!= null` since some options can be `null`.

cspotcode replied:
Do we want a `null` to be able to override a non-null value specified elsewhere?  The idea was `undefined` means the option is unspecified, whereas `null` means we explicitly want to override the default with `null`.  I was mimicking the JS language behavior of defaults in arguments and destructuring.  Will that cause problems?

blakeembrey replied:
> Will that cause problems?

No, I think your reasoning sounds solid. I hadn't considered someone would use `null`, it's historically just an input because of using the `yn` library which returns `null` when a value is neither true or false.

cspotcode commented:
Thanks for the speedy review.  I started working on the necessary changes before realizing I won't be able to get to this till at least next week.  I pushed what I have, and I'll do the rest as soon as I can.

blakeembrey commented:
@cspotcode I can also take a stab at landing the PR this weekend too. Since you've done the core functionality, I'd really be fixing all my own nit-picking since I'm particular about the structure 😄

cspotcode commented:
@blakeembrey That is totally fine with me, if you've got the free time.  Thanks!

cspotcode commented:
@blakeembrey This is ready for another review.  I think I've fixed all the issues:

* No more `OptionsHelper`; defaults are applied within `create()` and `readConfig()` as needed.
* Fixed inconsistent usage of `readFile` and `fileExists` in `readConfig()`
* Moved `"requires"` support into a follow-up PR
* Upgraded `yn` so that unspecified env vars will be `undefined` to be compatible with `defaults()` and JS semantics
* Fixed `files` handling in `readConfig`

cspotcode commented:
@blakeembrey bump!  If you're still on holiday, no rush, but when you have time I'd love to get this feature released.  Thanks in advance.

cspotcode commented:
@blakeembrey I hope it's ok if I bump this again.  Is there anything I can do to make this easier to review, rebasing or adding tests or whatever?  I'd love to get it into a release.  Anyone else I should request a review from if you're busy?

blakeembrey commented:
@cspotcode Taking a look now. Tried to have a proper vacation 😄 In the meantime I'll also add you as a collaborator in GitHub.

cspotcode commented:
@blakeembrey Thanks!  I didn't mean to interrupt your vacation if it's still going.

blakeembrey commented:
@cspotcode Nope, it's over. I intended to review this now.

blakeembrey commented on src/index.ts:
```diff
@@ -336,37 +336,32 @@ export function register (opts: RegisterOptions = {}): Register {
+  let { compiler, ts } = loadCompiler(inputOptions.compiler)
 
-  // Read config file
-  const { config, options: tsconfigOptions } = readConfig(cwd, ts, optionsWithoutDefaults)
```
Looks like the usage of `optionsWithoutDefaults` here is what broke https://github.com/TypeStrong/ts-node/issues/938. Fixing and adding a test now.
